File: src/main/java/com/github/svegon/capi/event/render/tooltip/AppendItemTooltipListener.java
package com.github.svegon.capi.event.render.tooltip;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.item.TooltipContext;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.item.ItemStack;
import net.minecraft.text.Text;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public interface AppendItemTooltipListener {
    Event<AppendItemTooltipListener> EVENT = EventFactory.createArrayBacked(AppendItemTooltipListener.class,
            (stack, player, context, tooltip) -> {}, listeners -> (stack, player, context, tooltip) -> {
                for (AppendItemTooltipListener listener : listeners) {
                    listener.appendItemTooltip(stack, player, context, tooltip);
                }
            });

    void appendItemTooltip(ItemStack stack, @Nullable PlayerEntity player, TooltipContext context, List<Text> tooltip);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/tooltip/AppendItemTooltipListener.kt
package io.github.svegon.capi.event.render.tooltip

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.item.TooltipContext
import net.minecraft.text.Text
import java.util.function.Function

interface AppendItemTooltipListener {
    fun appendItemTooltip(stack: ItemStack?, player: PlayerEntity?, context: TooltipContext?, tooltip: List<Text?>?)

    companion object {
        @JvmField
        val EVENT: Event<AppendItemTooltipListener?> = EventFactory.createArrayBacked<AppendItemTooltipListener?>(
            AppendItemTooltipListener::class.java,
            AppendItemTooltipListener { stack: ItemStack?, player: PlayerEntity?, context: TooltipContext?, tooltip: List<Text?>? -> },
            Function<Array<AppendItemTooltipListener?>, AppendItemTooltipListener?> { listeners: Array<AppendItemTooltipListener?>? ->
                AppendItemTooltipListener { stack: ItemStack?, player: PlayerEntity?, context: TooltipContext?, tooltip: List<Text?>? ->
                    for (listener in listeners) {
                        listener.appendItemTooltip(stack, player, context, tooltip)
                    }
                }
            })
    }
}
